### DISCUSSION TITLE
How to Set up NPM to Run with Webmin + Apache

### DISCUSSION THREAD

**Discussion by afunworm11** [NginxProxyManager/nginx-proxy-manager#3221] [Q&A]
Hi.

I have an Ubuntu server running webmin at :10000, apache running at 80 (came default when I installed webmin).

Then I have a docker container running NPM, but it runs on port 80, 443 and 81. I am wondering if there is any documentation on how to set up so that I can still run NPM with Apache and webmin.

Thank you.

**Comment by lastsamurai26** [comment]
80:80 -> port:80
change the first number 

example 8888:80 and so on

**Reply by afunworm11** [reply]
Thank you for your response.

However, isn't it that NPM should be listening to :80? And somehow I need to reconfigure Apache to listen to different ports to work with NPM?

Let's say I have app1.domain.com, app2.domain.com and web.domain.com, they are all pointing to my server's external IP.

I would like to access the internal app1, say localhost:1111 through app1.domain.com; localhost:2222 through app2.domain.com. For this to work, I need to use NPM that listens to port 80 and 443, right?

At the same time, I need to access web.domain.com through Apache which is currently listening to 80 and 443 also.

How do I achieve such thing?

My experience in networking is limited. I highly appreciate your help!

**Reply by lastsamurai26** [reply]
I have the same consterlation on another machine.
Docker 
-> NginxProxyManager
-> Portainer
-> Vaultwarden
-> Onlyoffice
-> Redis

Local 
-> Apache
-> Nextcloud
-> shlink

Unfortunately you have to give other ports to the local apache
example 8888 (http) and 4433 (https)

but still give apache all domains which should be local or the nginxproxymanager then just do the following

Https domain.com 4433

that works :-) and works for me without problems

**Reply by afunworm** [reply]
Hi.

Thanks again for the response. I think I am getting really close to resolving it. So now, I have:

- Apache listens on 4431 and 8011.
- NPM listens on 80 and 443.
- Proxy hosts that forwarded app1.domain.com, app2.domain.com, etc. to :8011 and :4431 (Apache).

The problem I am running into is that (I assume) even the proxy works, Apache doesn't know what domain is sending the request, so it cannot map it to the correct web directory. Visiting app1.domain.com and app2.domain.com results in Apache display the same (same location locally) default website.

Is there anything in advanced configuration I can use to forward the domain to the web server so it can map the correct location?

Thank you again!

**Reply by lastsamurai26** [reply]
You need to create for each domain a vhosts on the apache.
So NginxProxyManager accept the domain and will send the domain to the apache via port 4433 and the apache will accept it aswell because the vhosts settings

**Reply by afunworm** [reply]
Hi.

I'm sorry. My experience in this is pretty limited. Do I need to change anything from NPM, or vhosts from Apache is enough?

Would you be able to help me whip up a simple pseudo vhosts file? Or let me know what keyword to research to make this happen? I'm not sure what to put for the vhosts from on Apache.

Thank you!

**Reply by lastsamurai26** [reply]
Http
```
<virtualhost *:8011>
      # Admin email, Server Name (domain name) and any aliases
      ServerAdmin admin
      ServerName  domain.de
      ServerAlias domain.de
      ServerAlias www.domain.de
      DocumentRoot "/var/www/vhosts/domain.de/httpdocs"
      DirectoryIndex index.htm index.html index.php
        <Directory />
        Options FollowSymLinks
        AllowOverride None
        </Directory>
        <Directory /var/www/vhosts/domain.de/httpdocs>
        Options +Indexes +MultiViews +FollowSymLinks +ExecCGI
        Require all granted
           RewriteEngine On
           RewriteCond %{HTTPS} off
           RewriteRule (.*) https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301]
        AllowOverride all
        </Directory>
      LogLevel warn
      ErrorLog  /var/www/vhosts/domain.de/logs/error.log
      CustomLog /var/www/vhosts/domain.de/logs/access.log combined
      ServerSignature On
</virtualhost>
```


Https
```
<IfModule mod_ssl.c>
<VirtualHost *:4431>
      # Admin email, Server Name (domain name) and any aliases
      ServerAdmin admin
      ServerName  domain.de
      ServerAlias domain.de
      ServerAlias www.domain.de
      Header always set Strict-Transport-Security "max-age=31536000; includeSubdomains;"
      Header always set X-XSS-Protection "1; mode=block"
      Header always set Referrer-Policy "strict-origin-when-cross-origin"
      Header always set X-Frame-Options SAMEORIGIN
      Header always edit Set-Cookie ^(.*)$ $1;HttpOnly;Secure
      Header always unset X-Content-Type-Options
      Header always set Content-Security-Policy "frame-ancestors 'self'"

      # Optionals
      # RequestHeader unset X-Forwarded-Host

      # Remove Drupal Headers
      Header always unset X-Generator
      Header always unset X-Drupal-Cache
      Header always unset X-Drupal-Dynamic-Cache
      Header always unset X-Powered-By
      Protocols h2 http/1.1
      DocumentRoot "/var/www/vhosts/domain.de/httpdocs"
      DirectoryIndex index.htm index.html index.php
      SSLEngine on

        <FilesMatch "\.(cgi|shtml|phtml|php)$">
            SSLOptions +StdEnvVars
        </FilesMatch>
        <Directory /usr/lib/cgi-bin>
           SSLOptions +StdEnvVars
        </Directory>
        <Directory />
        Options FollowSymLinks
        AllowOverride None
        </Directory>

        <Directory /var/www/vhosts/domain.de/httpdocs/>
        Options +Indexes +MultiViews +FollowSymLinks +ExecCGI
        Require all granted
        AllowOverride all
        </Directory>

      LogLevel warn
      ErrorLog  /var/www/vhosts/domain.de/logs/error.log
      CustomLog /var/www/vhosts/domain.de/logs/access.log combined
      ServerSignature On
      Include /etc/letsencrypt/options-ssl-apache.conf
      SSLCertificateFile /etc/letsencrypt/live/domain.de/fullchain.pem
      SSLCertificateKeyFile /etc/letsencrypt/live/domain.de/privkey.pem

</virtualhost>
</IfModule>
```


But you can't use this by default because i don't know your setup :-( 
But this is mean as example

**Reply by lastsamurai26** [reply]
I hope this helps

**Reply by afunworm** [reply]
Thank you so much!!! I just didn't know what to look for or to Google. I shall modify this and hopefully it'll work!

**Reply by afunworm** [reply]
So I've tried it. The server already have vhosts configured for each domain app1.domain.com, app2.domain.com, etc. Do you know which line in your example is import?

Right now mine is looking like this:

```
ServerName app1
DocumentRoot /home/domain/subdomains/app1/public_html
ErrorLog /var/log/virtualmin/polls_error_log
CustomLog /var/log/virtualmin/app1_access_log combined
ScriptAlias /cgi-bin/ /home/domain/subdomains/app1/cgi-bin/
ScriptAlias /awstats/ /home/domain/subdomains/app1/cgi-bin/
DirectoryIndex index.php index.php4 index.php5 index.htm index.html
<Directory /home/domain/subdomains/poll/public_html>
    Options -Indexes +IncludesNOEXEC +SymLinksIfOwnerMatch 
    Require all granted
    AllowOverride All Options=ExecCGI,Includes,IncludesNOEXEC,Indexes,MultiViews,SymLinksIfOwnerMatch
    AddType text/plain .php
</Directory>
<Directory /home/domain/subdomains/app1/cgi-bin>
    Require all granted
    AllowOverride All Options=ExecCGI,Includes,IncludesNOEXEC,Indexes,MultiViews,SymLinksIfOwnerMatch
    SetHandler proxy:unix:/var/fcgiwrap/169597014916980.sock/socket|fcgi://localhost
    ProxyFCGIBackendType GENERIC
</Directory>
RemoveHandler .php
```
I just don't know what line is important from your example to put into my vhosts. Also, would you please enlighten me on how changing these vhosts would help? I believe NginxProxyManager, after forwarding the traffic to the correct internal IP for Apache, Apache just doesn't know what the hostname/domain that is requesting the data is, so it can't map it to the write home directory.

Is there anything I can do in NginxProxyManager's advance section that will forward the hostname to :8011 instead of just redirecting?

Thank you for trying to help me out with this.

**Reply by lastsamurai26** [reply]
Maybe it is better to compare my vhost and your to find what is different. 
It may good to find the difference yourself :-) if not i will make one of the vhosts from you for you. Possibly later in the day or in a week depending on how much time I have.

The apache needs to know each domain which he is hosting the nginxproxymanager reverse only the request and send it to https://domain.de:4431

For now the first 2 lines of my vhosts id important.
